c++/src/kj/compat/http.c++

```diff
@@ -6310,32 +6370,59 @@ public:
   kj::Promise<void> connect(kj::StringPtr host, const HttpHeaders& headers,
                             ConnectResponse& tunnel) override {
     KJ_REQUIRE(!headers.isWebSocket(), "WebSocket upgrade headers are not permitted in a connect.");
+
     return client.connect(host, headers)
         .then([&tunnel](HttpClient::ConnectResponse&& response) -> kj::Promise<void> {
-      KJ_SWITCH_ONEOF(response.connectionOrBody) {
-        KJ_CASE_ONEOF(io, kj::Own<kj::AsyncIoStream>) {
-          auto io2 = tunnel.accept(response.statusCode, response.statusText, *response.headers);
-          auto promises = kj::heapArrayBuilder<kj::Promise<void>>(2);
-
-          promises.add(io->pumpTo(*io2).then([&io2 = *io2](uint64_t size) {
-            io2.shutdownWrite();
-          }).eagerlyEvaluate(nullptr));
-          promises.add(io2->pumpTo(*io).then([&io = *io](uint64_t size) {
-            io.shutdownWrite();
-          }).eagerlyEvaluate(nullptr));
-          return kj::joinPromises(promises.finish()).attach(kj::mv(io), kj::mv(io2));
-        }
-        KJ_CASE_ONEOF(input, kj::Own<kj::AsyncInputStream>) {
-          auto io2 = tunnel.reject(response.statusCode, response.statusText, *response.headers);
-          return input->pumpTo(*io2).ignoreResult().attach(kj::mv(input), kj::mv(io2));
-        }
-      }
-      KJ_UNREACHABLE;
+      auto maybeCanceler = heap<MaybeCanceler>(kj::mv(response.status));
+
+      // This operates optimistically. In order to support pipelining, we have to
+      // accept the tunnel early, even if we're not yet certain that the tunnel can
+      // actually be established. If the response.status promise rejects or returns
+      // a failure, then we will interrupt the pipelines we establish below.
+      auto io = kj::mv(response.connection);
+      HttpHeaderTable headerTable {};
+      HttpHeaders responseHeaders(headerTable);
+      auto io2 = tunnel.accept(200, "OK"_kj, responseHeaders);
```

Comment:
Isn't it a problem that you're always sending empty response headers here?

I think we need to change `HttpService::connect()` to support pipelining. It would need to take an `AsyncIoStream&` representing the connection as a parameter. We would document that `tunnel.accept()` must be called before writing to this stream (or, perhaps, that writes will be blocked until `accept()` has been called).

Reply:
The empty headers is certainly not ideal but I wouldn't call it a problem. We currently don't do anything with the response headers. That said, I'm not all that happy with this particular bit yet.

Reply:
Your specific use case might not care about the headers, but we're implementing a general library here. If response headers can't be produced correctly then the API should be changed so that it doesn't pretend to produce them at all. But I think passing in the stream to `connect()` will end up with cleaner code anyway, e.g. implementing the HttpClient adapter should be easier if it doesn't have to wait for the service to call `accept()`.